### PR TITLE
Support for Clear Linux

### PR DESCRIPTION
**Pull request by nathanchance**
Clear Linux is rather odd in a few places (see the individual commit messages for the full details). I feel this distribution is starting to grow so I think it is worth supporting here.

cc @kdrag0n, just in case there is anything here that might break Proton Clang.

This has been run through my `podman` testing infrastructure ([link to current version](https://github.com/nathanchance/scripts/blob/991930822f98e153ebbfad6a12c7c170e2439a49/env/stubs/cbl#L112-L312)), where it passes for all distributions still.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -88,6 +88,39 @@ These scripts have been tested in a Docker image of the following distributions,
...
+  $ curl -LSs https://ftp.denx.de/pub/u-boot/u-boot-2021.01.tar.bz2 | tar -xjf -
+  $ cd u-boot-2021.01
+  $ make -j"$(nproc)" defconfig tools-all
+  ...
+  $ sudo install -Dm755 tools/mkimage /usr/local/bin/mkimage
+  $ mkimage -V
+  mkimage version 2021.01
+  ```
+
+  Lastly, Clear Linux has `${CC}`, `${CXX}`, `${CFLAGS}`, and `${CXXFLAGS}` in the environment, which messes with the heuristics of the script for selecting a compiler. By default, the script will attempt to use `clang` and `ld.lld` but the environment's value of `${CC}` and `${CXX}` is respected first so `gccc` and `g++` will be used. Clear Linux has optimized their `gcc` and `g++` so this is fine but if you would like to use `clang` and `clang++` instead, invoke the script like so:
```
Comment (msfjarvis) [inline README.md:117]:
```suggestion
  Lastly, Clear Linux has `${CC}`, `${CXX}`, `${CFLAGS}`, and `${CXXFLAGS}` in the environment, which messes with the heuristics of the script for selecting a compiler. By default, the script will attempt to use `clang` and `ld.lld` but the environment's value of `${CC}` and `${CXX}` is respected first so `gcc` and `g++` will be used. Clear Linux has optimized their `gcc` and `g++` so this is fine but if you would like to use `clang` and `clang++` instead, invoke the script like so:
```

**Review by msfjarvis — APPROVED** [review]
LGTM, there's also a typo in 4c51b5a3fb1d1d69dee89627d2d23800f5af575f's commit message (`s/camke/cmake/`) if you wanna fix that.

**Comment by nathanchance** [comment]
Thanks for the review, I fixed up both issues. I will commit this in 12 hours or so if there are no objections.

**Comment by dileks** [comment]
Can Debian's `LLVM_HOST_TRIPLE` be fixed, too?
```sh
$ grep warning: build-log_5.12.0-rc1-9-amd64-clang13-cfi.txt 
26:dpkg-architecture: warning: specified GNU system type x86_64-linux-gnu does not match CC system type x86_64-unknown-linux-gnu, try setting a correct CC environment variable
```

**Comment by nathanchance** [comment]
I am not familiar with Debian's packing infrastructure or how to easily determine what the correct triple should be for Debian's case.

For Clear Linux, it is obvious because Clear Linux only supports one architecture (x86_64) and the cmake option is right [here](https://github.com/clearlinux-pkgs/llvm/blob/383e7d1ff0a3fc9ddf79cf2efeef9f6e5d5139f7/cmake_args#L39) plus the build errors without that.

For Debian, they only appear to set `LLVM_HOST_TRIPLE` for a couple of architectures ([source](https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/blob/11/debian/rules#L98)). You are welcome to use `-D LLVM_HOST_TRIPLE=...` to try and fix that warning and if it can be written in a portable way, imported into the script.

**Review by kdrag0n — APPROVED** [review]
This doesn't break anything in my workflow.

**Comment by dileks** [comment]
I have added to my build-script:
```sh
### LLVM_HOST_TRIPLE options
# Debian: dpkg-architecture: warning:
# specified GNU system type x86_64-linux-gnu does not match CC system type x86_64-unknown-linux-gnu, try setting a correct CC environment variable
# This changes in CMakeCache.txt:
# 1. LLVM_DEFAULT_TARGET_TRIPLE:STRING=x86_64-linux-gnu
# 2. LLVM_HOST_TRIPLE:STRING=x86_64-linux-gnu
DEBIAN_GNU_SYSTEM_TYPE="x86_64-linux-gnu"
LLVM_HOST_TRIPLE_OPTS="-D LLVM_HOST_TRIPLE=$DEBIAN_GNU_SYSTEM_TYPE"
```

I did a ccached-rebuild .
`llc` binary shows the new target `x86_64-linux-gnu` but `clang` still shows `x86_64-unknown-linux-gnu` with `--version` option.
Unfortunately, the warning in the Linux-kernel build is still there.
I will try to do a full (with cleaned ccache's cache) new LLVM toolchain build with a new upgrade.

**Comment by dileks** [comment]
NOPE.
```sh
$ /opt/llvm-toolchain/bin/llc --version
LLVM (http://llvm.org/):
  LLVM version 13.0.0git
  Optimized build.
  Default target: x86_64-linux-gnu
  Host CPU: sandybridge

  Registered Targets:
    bpf    - BPF (host endian)
    bpfeb  - BPF (big endian)
    bpfel  - BPF (little endian)
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64

$ /opt/llvm-toolchain/bin/clang --version
dileks clang version 13.0.0 (https://github.com/llvm/llvm-project.git 9c0274cdeae904089806be6faee72b9126d2cf5b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/llvm-toolchain/bin

$ /opt/llvm-toolchain/bin/llvm-dwarfdump --version
LLVM (http://llvm.org/):
  LLVM version 13.0.0git
  Optimized build.
  Default target: x86_64-linux-gnu
  Host CPU: sandybridge
```

File: build-llvm.py [hunk 2]
```diff
@@ -759,6 +750,22 @@ def cc_ld_cmake_defines(dirs, env_vars, stage):
...
+    """
+    Generate distribution specific cmake defines
+    :return: A set of defines
+    """
+    defines = {}
+
+    # Clear Linux needs a different target to find all of the C++ header files, otherwise
+    # stage 2+ compiles will fail without this
+    # We figure this out based on the existence of x86_64-generic-linux in the C++ headers path
+    if glob.glob("/usr/include/c++/*/x86_64-generic-linux"):
```
Comment (MaskRay) [inline build-llvm.py:763]: Does `/usr/x86_64-generic-linux/include` exist? If yes, it should be preferred.

**Pull request merged**